glom/matching.py

```diff
@@ -0,0 +1,327 @@
+        else:
+            and_ = And(self, other)
+        and_.chain = True
+        return and_
+
+    __rand__ = __and__
+
+    def __or__(self, other):
+        if self is M:
+            return Or(other)
+        return Or(self, other)
+
+    def glomit(self, target, scope):
+        lhs, op, rhs = self.lhs, self.op, self.rhs
+        if lhs is M:
+            lhs = target
+        if rhs is M:
+            rhs = target
+        matched = (
+            (op == '=' and lhs == rhs) or
+            (op == '!' and lhs != rhs) or
+            (op == '>' and lhs > rhs) or
+            (op == '<' and lhs < rhs) or
+            (op == 'g' and lhs >= rhs) or
+            (op == 'l' and lhs <= rhs)
+        )
+        if matched:
+            return target
+        raise GlomMatchError("{!r} {} {!r}", lhs, _M_OP_MAP.get(op, op), rhs)
+
+    def __repr__(self):
+        if self is M:
+            return "M"
+        op = _M_OP_MAP.get(self.op, self.op)
+        return "{!r} {} {!r}".format(self.lhs, op, self.rhs)
+
+
+M = MType(None, None, None)
+
+
+_MISSING = make_sentinel('MISSING')
+
+
+class Optional(object):
+    """
+    mark a key as optional in a dictionary
+
+    by default all non-exact-match type keys are optional
+    """
+    __slots__ = ('key',)
+
+    def __init__(self, key):
+        assert _precedence(key) == 0, "key must be == match"
+        self.key = key
+
+    def glomit(self, target, scope):
+        if target != self.key:
+            raise GlomMatchError("target {} != spec {}", target, self.key)
+
+    def __repr__(self):
+        return 'Optional({!r})'.format(self.key)
+
+
+class Required(object):
+    """
+    mark a key as required in a dictionary
+
+    by default, only exact-match type keys are required
+    """
+    __slots__ = ('key',)
+
+    def __init__(self, key):
+        assert _precedence(key) != 0, "== match keys are already required"
+        self.key = key
+
+    def glomit(self, target, scope):
+        return scope[glom](target, self.key, scope)
+
+    def __repr__(self):
+        return 'Required({!r})'.format(self.key)
+
+
+def _precedence(match):
+    """
+    in a dict spec, target-keys may match many
+    spec-keys (e.g. 1 will match int, M > 0, and 1);
+    therefore we need a precedence for which order to try
+    keys in; higher = later
+    """
+    if type(match) in (Required, Optional):
+        match = match.key
+    if match is DEFAULT:
+        return 5
+    if type(match) in (list, tuple, set, frozenset, dict):
+        return 4
+    if isinstance(match, type):
+        return 3
+    if hasattr(match, "glomit"):
+        return 2
+    if callable(match):
+        return 1
+    return 0  # == match
+
+
+def _handle_dict(target, spec, scope):
+    if not isinstance(target, dict):
+        raise GlomTypeMatchError(type(target), dict)
+    spec_keys = spec  # cheating a little bit here, list-vs-dict, but saves an object copy sometimes
+    if sys.version_info < (3, 6):
```

Comment:
(We could also run our own test, throw a few dozen items into a dict at import time, see if they come out the same order?)

Reply:
I'm a little nervous about testing ordering accidentally working on certain platforms, but 100% agree we should use a good "is_sorted_dict"

hey!  does that sound like a bolton to you?  `dict_utils.DICT_IS_ORDERED`?

Reply:
Sure, it can go into boltons. Also, a hundred random items going into a dict and coming out the same exact order, that'd be mighty strange yeah?

Reply:
I could imagine some implementation deciding that linear search is faster than a real table maybe?  (since it is mostly going to be doing integer comparisons on a hash)

I guess it will become less and less relevant what mechanism we use over time

Reply:
(since I expect sorted to become more and more the default)